// @newrelic/browser-agent/dist/cjs/features/logging/aggregate/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Aggregate = void 0;
var _registerHandler = require("../../../common/event-emitter/register-handler");
var _console = require("../../../common/util/console");
var _stringify = require("../../../common/util/stringify");
var _aggregateBase = require("../../utils/aggregate-base");
var _constants = require("../constants");
var _log = require("../shared/log");
var _utils = require("../shared/utils");
var _traverse = require("../../../common/util/traverse");
var _constants2 = require("../../../common/session/constants");
var _constants3 = require("../../session_replay/constants");
var _featureGates = require("../../utils/feature-gates");
var _utils2 = require("../../../common/v2/utils");
/**
 * Copyright 2020-2026 New Relic, Inc. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

const LOGGING_EVENT = 'Logging/Event/';
class Aggregate extends _aggregateBase.AggregateBase {
  static featureName = _constants.FEATURE_NAME;
  constructor(agentRef) {
    super(agentRef, _constants.FEATURE_NAME);
    const updateLocalLoggingMode = (auto, api) => {
      this.loggingMode = {
        auto,
        api
      };
      // In agent v1.290.0 & under, the logApiMode prop did not yet exist, so need to account for old session state being in-use.
      if (api === undefined) this.loggingMode.api = auto;
    };

    /** set up agg-level behaviors specific to this feature */
    this.harvestOpts.raw = true;
    super.customAttributesAreSeparate = true;

    // The SessionEntity class can emit a message indicating the session was cleared and reset (expiry, inactivity). This feature must abort and never resume if that occurs.
    this.ee.on(_constants2.SESSION_EVENTS.RESET, () => {
      this.abort(_constants3.ABORT_REASONS.RESET);
    });
    this.ee.on(_constants2.SESSION_EVENTS.UPDATE, (type, data) => {
      if (this.blocked || type !== _constants2.SESSION_EVENT_TYPES.CROSS_TAB) return;
      // In agent v1.290.0 & under, the logApiMode prop did not yet exist, so need to account for old session state being in-use with just loggingMode off == feature off.
      if (data.loggingMode === _constants.LOGGING_MODE.OFF && (!data.logApiMode || data.logApiMode === _constants.LOGGING_MODE.OFF)) this.abort(_constants3.ABORT_REASONS.CROSS_TAB);else updateLocalLoggingMode(data.loggingMode, data.logApiMode);
    });
    this.waitForFlags(['log', 'logapi']).then(([auto, api]) => {
      if (this.blocked) return; // means abort already happened before this, likely from session reset or update; abort would've set mode off + deregistered drain

      this.loggingMode ??= {
        auto,
        api
      }; // likewise, don't want to overwrite the mode if it was set already
      const session = this.agentRef.runtime.session;
      if ((0, _featureGates.canEnableSessionTracking)(agentRef.init) && session) {
        if (session.isNew) this.#syncWithSessionManager();else updateLocalLoggingMode(session.state.loggingMode, session.state.logApiMode);
      }
      if (this.loggingMode.auto === _constants.LOGGING_MODE.OFF && this.loggingMode.api === _constants.LOGGING_MODE.OFF) {
        this.blocked = true;
        this.deregisterDrain();
        return;
      }

      /** emitted by instrument class (wrapped loggers) or the api methods directly */
      (0, _registerHandler.registerHandler)(_constants.LOGGING_EVENT_EMITTER_CHANNEL, this.handleLog.bind(this), this.featureName, this.ee);
      this.drain();
      /** harvest immediately once started to purge pre-load logs collected */
      agentRef.runtime.harvester.triggerHarvestFor(this);
    });
  }
  handleLog(timestamp, message, attributes = {}, level = _constants.LOG_LEVELS.INFO, autoCaptured, target) {
    if (this.blocked) return;
    // Check respective logging mode depending on whether this log is from auto wrapped instrumentation or manual API that it's not turned off.
    const modeForThisLog = autoCaptured ? this.loggingMode.auto : this.loggingMode.api;
    if (!modeForThisLog) return;
    if (!attributes || typeof attributes !== 'object') attributes = {};
    if (typeof level === 'string') level = level.toUpperCase();
    if (!(0, _utils.isValidLogLevel)(level)) return (0, _console.warn)(30, level);
    if (modeForThisLog < (_constants.LOGGING_MODE[level] || Infinity)) {
      this.reportSupportabilityMetric(LOGGING_EVENT + 'Dropped/Sampling');
      return;
    }
    try {
      if (typeof message !== 'string') {
        const stringified = (0, _stringify.stringify)(message);
        /**
           * Error instances convert to `{}` when stringified
           * Symbol converts to '' when stringified
           * other cases tbd
           * */
        if (!!stringified && stringified !== '{}') message = stringified;else message = String(message);
      }
    } catch (err) {
      (0, _console.warn)(16, message);
      this.reportSupportabilityMetric(LOGGING_EVENT + 'Dropped/Casting');
      return;
    }
    if (typeof message !== 'string' || !message) return (0, _console.warn)(32);
    const addEvent = attributes => {
      const log = new _log.Log(Math.floor(this.agentRef.runtime.timeKeeper.correctRelativeTimestamp(timestamp)), message, attributes, level);
      if (this.events.add(log)) this.reportSupportabilityMetric(LOGGING_EVENT + (autoCaptured ? 'Auto' : 'API') + '/Added');
    };
    addEvent({
      ...attributes,
      ...(0, _utils2.getVersion2Attributes)(target, this)
    });
    if ((0, _utils2.shouldDuplicate)(target, this)) addEvent({
      ...attributes,
      ...(0, _utils2.getVersion2DuplicationAttributes)(target, this)
    });
  }
  serializer(eventBuffer) {
    const sessionEntity = this.agentRef.runtime.session;
    return [{
      common: {
        /** Attributes in the `common` section are added to `all` logs generated in the payload */
        attributes: {
          ...(0, _traverse.applyFnToProps)(this.agentRef.info.jsAttributes, this.obfuscator.obfuscateString.bind(this.obfuscator), 'string'),
          ...(this.harvestEndpointVersion === 1 && {
            'entity.guid': this.agentRef.runtime.appMetadata.agents[0].entityGuid,
            appId: this.agentRef.info.applicationID
          }),
          ...(sessionEntity && {
            session: sessionEntity.state.value || '0',
            // The session ID that we generate and keep across page loads
            hasReplay: sessionEntity.state.sessionReplayMode === 1,
            // True if a session replay recording is running
            hasTrace: sessionEntity.state.sessionTraceMode === 1 // True if a session trace recording is running
          }),
          ptid: this.agentRef.runtime.ptid,
          // page trace id
          standalone: Boolean(this.agentRef.info.sa),
          // copy paste (true) vs APM (false)
          agentVersion: this.agentRef.runtime.version,
          // browser agent version
          // The following 3 attributes are evaluated and dropped at ingest processing time and do not get stored on NRDB:
          'instrumentation.provider': 'browser',
          'instrumentation.version': this.agentRef.runtime.version,
          'instrumentation.name': this.agentRef.runtime.loaderType
        }
      },
      /** logs section contains individual unique log entries */
      logs: (0, _traverse.applyFnToProps)(eventBuffer, this.obfuscator.obfuscateString.bind(this.obfuscator), 'string')
    }];
  }
  queryStringsBuilder() {
    return {
      browser_monitoring_key: this.agentRef.info.licenseKey
    };
  }

  /** Abort the feature, once aborted it will not resume */
  abort(reason = {}) {
    this.reportSupportabilityMetric("Logging/Abort/".concat(reason.sm));
    this.blocked = true;
    if (this.events) {
      this.events.clear();
      this.events.clearSave();
    }
    this.loggingMode = {
      auto: _constants.LOGGING_MODE.OFF,
      api: _constants.LOGGING_MODE.OFF
    };
    this.#syncWithSessionManager();
    this.deregisterDrain();
  }
  #syncWithSessionManager() {
    this.agentRef.runtime.session?.write({
      loggingMode: this.loggingMode.auto,
      logApiMode: this.loggingMode.api
    });
  }
}
exports.Aggregate = Aggregate;
