// @newrelic/browser-agent/dist/cjs/features/utils/agent-session.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setupAgentSession = setupAgentSession;
var _drain = require("../../common/drain/drain");
var _registerHandler = require("../../common/event-emitter/register-handler");
var _sessionEntity = require("../../common/session/session-entity");
var _localStorage = require("../../common/storage/local-storage.js");
var _constants = require("../../common/session/constants");
var _info = require("../../common/config/info");
var _attributeSize = require("../../common/util/attribute-size");
var _handle = require("../../common/event-emitter/handle");
var _constants2 = require("../../loaders/api/constants");
var _constants3 = require("../metrics/constants");
var _features = require("../../loaders/features/features");
var _sharedHandlers = require("../../loaders/api/sharedHandlers");
/**
 * Copyright 2020-2026 New Relic, Inc. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

function setupAgentSession(agentRef) {
  if (agentRef.runtime.session) return agentRef.runtime.session; // already setup

  const sessionInit = agentRef.init.session;
  agentRef.runtime.session = new _sessionEntity.SessionEntity({
    agentRef,
    key: _constants.DEFAULT_KEY,
    storage: new _localStorage.LocalStorage(),
    expiresMs: sessionInit?.expiresMs,
    inactiveMs: sessionInit?.inactiveMs
  });

  // Retrieve & re-add all of the persisted setCustomAttribute|setUserId k-v from previous page load(s), if any was stored.
  const customSessionData = agentRef.runtime.session.state.custom;
  if (customSessionData && Object.keys(customSessionData).length) {
    /** stored attributes from previous page should not take precedence over attributes stored on this page via API before the page load */
    agentRef.info = (0, _info.mergeInfo)({
      ...agentRef.info,
      jsAttributes: {
        ...customSessionData,
        ...agentRef.info.jsAttributes
      }
    });
  }

  /** track changes to the jsAttributes field over time for aiding with harvest mechanics */
  agentRef.runtime.jsAttributesMetadata = (0, _attributeSize.trackObjectAttributeSize)(agentRef.info, 'jsAttributes');
  const sharedEE = agentRef.ee;

  // any calls to newrelic.setCustomAttribute(<persisted>) will need to be added to:
  // local info.jsAttributes {}
  // the session's storage API
  (0, _registerHandler.registerHandler)('api-setCustomAttribute', (time, key, value) => {
    agentRef.runtime.session.syncCustomAttribute(key, value);
  }, 'session', sharedEE);

  // any calls to newrelic.setUserId(...) will need to be added to:
  // local info.jsAttributes {}
  // the session's storage API
  (0, _registerHandler.registerHandler)('api-setUserId', (time, key, value) => {
    agentRef.runtime.session.syncCustomAttribute(key, value);
  }, 'session', sharedEE);
  (0, _registerHandler.registerHandler)('api-setUserIdAndResetSession', value => {
    agentRef.runtime.session.reset();
    (0, _handle.handle)(_constants3.SUPPORTABILITY_METRIC_CHANNEL, ['API/' + _constants2.SET_USER_ID + '/resetSession/called'], undefined, _features.FEATURE_NAMES.metrics, sharedEE);
    (0, _sharedHandlers.appendJsAttribute)(agentRef, 'enduser.id', value, _constants2.SET_USER_ID, true);
  }, 'session', sharedEE);
  (0, _registerHandler.registerHandler)('api-consent', accept => {
    agentRef.runtime.session.write({
      consent: accept === undefined ? true : accept
    });
  }, 'session', sharedEE);
  (0, _drain.drain)(agentRef, 'session');
  return agentRef.runtime.session;
}
