// @newrelic/browser-agent/dist/cjs/loaders/configure/configure.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.configure = configure;
var _topLevelCallers = require("../api/topLevelCallers");
var _nreum = require("../../common/window/nreum");
var _info = require("../../common/config/info");
var _init = require("../../common/config/init");
var _runtime = require("../../common/config/runtime");
var _runtime2 = require("../../common/constants/runtime");
var _publicPath = require("./public-path");
var _contextualEe = require("../../common/event-emitter/contextual-ee");
var _globalEvent = require("../../common/dispatch/global-event");
var _loaderConfig = require("../../common/config/loader-config");
/**
 * Copyright 2020-2026 New Relic, Inc. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * Sets or re-sets the agent's configuration values from global settings. This also attach those as properties to the agent instance.
 * IMPORTANT: setNREUMInitializedAgent must be called on the agent prior to calling this function.
 */
function configure(agent, opts = {}, loaderType, forceDrain) {
  // eslint-disable-next-line camelcase
  let {
    init,
    info,
    loader_config,
    runtime = {},
    exposed = true
  } = opts;
  if (!info) {
    const nr = (0, _nreum.gosCDN)();
    init = nr.init;
    info = nr.info;
    // eslint-disable-next-line camelcase
    loader_config = nr.loader_config;
  }
  agent.init = (0, _init.mergeInit)(init || {});
  // eslint-disable-next-line camelcase
  agent.loader_config = (0, _loaderConfig.mergeLoaderConfig)(loader_config || {});
  info.jsAttributes ??= {};
  if (_runtime2.isWorkerScope) {
    // add a default attr to all worker payloads
    info.jsAttributes.isWorker = true;
  }
  agent.info = (0, _info.mergeInfo)(info);
  const updatedInit = agent.init;
  agent.runtime ??= (0, _runtime.mergeRuntime)(runtime);

  // Apply proxy settings whenever configure is called (supports late {init} setting)
  if (updatedInit.proxy.assets) {
    (0, _publicPath.redefinePublicPath)(updatedInit.proxy.assets);
  }
  if (!agent.runtime.configured) {
    Object.defineProperty(agent, 'beacons', {
      get() {
        return [agent.info.beacon, agent.info.errorBeacon, agent.init.proxy.assets, agent.init.proxy.beacon].filter(Boolean);
      }
    });
    Object.defineProperty(agent.runtime, 'denyList', {
      get() {
        // Compute the internal traffic list fresh each time to ensure beacons array is current
        return [...(agent.init.ajax.deny_list || []), ...(agent.init.ajax.block_internal ? agent.beacons : [])];
      }
    });
    agent.runtime.ptid = agent.agentIdentifier;
    (0, _topLevelCallers.setTopLevelCallers)(agent); // no need to set global APIs on newrelic obj more than once

    agent.runtime.loaderType = loaderType;
    agent.ee = _contextualEe.ee.get(agent.agentIdentifier);
    agent.exposed = exposed;
    (0, _globalEvent.dispatchGlobalEvent)({
      drained: !!agent.runtime.activatedFeatures,
      type: 'lifecycle',
      name: 'initialize',
      feature: undefined,
      data: agent.config
    });
    agent.runtime.configured = true;
  }
}
