// @newrelic/browser-agent/dist/cjs/loaders/api/addToTrace.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setupAddToTraceAPI = setupAddToTraceAPI;
var _runtime = require("../../common/constants/runtime");
var _handle = require("../../common/event-emitter/handle");
var _console = require("../../common/util/console");
var _features = require("../features/features");
var _constants = require("./constants");
var _sharedHandlers = require("./sharedHandlers");
/**
 * Copyright 2020-2025 New Relic, Inc. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

function setupAddToTraceAPI(agent) {
  (0, _sharedHandlers.setupAPI)(_constants.ADD_TO_TRACE, function (evt) {
    if (!(evt && typeof evt === 'object' && evt.name && evt.start)) return;
    const report = {
      n: evt.name,
      s: evt.start - _runtime.originTime,
      e: (evt.end || evt.start) - _runtime.originTime,
      o: evt.origin || '',
      t: 'api'
    };
    if (report.s < 0 || report.e < 0 || report.e < report.s) {
      (0, _console.warn)(61, {
        start: report.s,
        end: report.e
      });
      return;
    }
    (0, _handle.handle)('bstApi', [report], undefined, _features.FEATURE_NAMES.sessionTrace, agent.ee);
  }, agent);
}
