// @newrelic/browser-agent/dist/cjs/loaders/api/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.spaPrefix = exports.prefix = exports.WRAP_LOGGER = exports.START = exports.SET_USER_ID = exports.SET_PAGE_VIEW_NAME = exports.SET_ERROR_HANDLER = exports.SET_CUSTOM_ATTRIBUTE = exports.SET_CURRENT_ROUTE_NAME = exports.SET_APPLICATION_VERSION = exports.REGISTER = exports.RECORD_REPLAY = exports.RECORD_CUSTOM_EVENT = exports.PAUSE_REPLAY = exports.NOTICE_ERROR = exports.MEASURE = exports.LOG = exports.INTERACTION = exports.FINISHED = exports.CONSENT = exports.ADD_TO_TRACE = exports.ADD_RELEASE = exports.ADD_PAGE_ACTION = void 0;
/**
 * Copyright 2020-2025 New Relic, Inc. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
const prefix = exports.prefix = 'api-';
const spaPrefix = exports.spaPrefix = prefix + 'ixn-';
const ADD_PAGE_ACTION = exports.ADD_PAGE_ACTION = 'addPageAction';
const ADD_TO_TRACE = exports.ADD_TO_TRACE = 'addToTrace';
const ADD_RELEASE = exports.ADD_RELEASE = 'addRelease';
const FINISHED = exports.FINISHED = 'finished';
const INTERACTION = exports.INTERACTION = 'interaction';
const LOG = exports.LOG = 'log';
const NOTICE_ERROR = exports.NOTICE_ERROR = 'noticeError';
const PAUSE_REPLAY = exports.PAUSE_REPLAY = 'pauseReplay';
const RECORD_CUSTOM_EVENT = exports.RECORD_CUSTOM_EVENT = 'recordCustomEvent';
const RECORD_REPLAY = exports.RECORD_REPLAY = 'recordReplay';
const REGISTER = exports.REGISTER = 'register';
const SET_APPLICATION_VERSION = exports.SET_APPLICATION_VERSION = 'setApplicationVersion';
const SET_CURRENT_ROUTE_NAME = exports.SET_CURRENT_ROUTE_NAME = 'setCurrentRouteName';
const SET_CUSTOM_ATTRIBUTE = exports.SET_CUSTOM_ATTRIBUTE = 'setCustomAttribute';
const SET_ERROR_HANDLER = exports.SET_ERROR_HANDLER = 'setErrorHandler';
const SET_PAGE_VIEW_NAME = exports.SET_PAGE_VIEW_NAME = 'setPageViewName';
const SET_USER_ID = exports.SET_USER_ID = 'setUserId';
const START = exports.START = 'start';
const WRAP_LOGGER = exports.WRAP_LOGGER = 'wrapLogger';
const MEASURE = exports.MEASURE = 'measure';
const CONSENT = exports.CONSENT = 'consent';
