// @newrelic/browser-agent/dist/cjs/features/soft_navigations/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.POPSTATE_TRIGGER = exports.POPSTATE_MERGE_WINDOW = exports.NO_LONG_TASK_WINDOW = exports.NODE_TYPE = exports.IPL_TRIGGER_NAME = exports.INTERACTION_TYPE = exports.INTERACTION_TRIGGERS = exports.INTERACTION_STATUS = exports.FEATURE_NAME = exports.API_TRIGGER_NAME = void 0;
var _features = require("../../loaders/features/features");
/**
 * Copyright 2020-2025 New Relic, Inc. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

const INTERACTION_TRIGGERS = exports.INTERACTION_TRIGGERS = ['click',
// e.g. user clicks link or the page back/forward buttons
'keydown',
// e.g. user presses left and right arrow key to switch between displayed photo gallery
'submit' // e.g. user clicks submit butotn or presses enter while editing a form field
];
const POPSTATE_TRIGGER = exports.POPSTATE_TRIGGER = 'popstate'; // e.g. user clicks browser back/forward button or history API is used programmatically
const API_TRIGGER_NAME = exports.API_TRIGGER_NAME = 'api';
const IPL_TRIGGER_NAME = exports.IPL_TRIGGER_NAME = 'initialPageLoad';
const FEATURE_NAME = exports.FEATURE_NAME = _features.FEATURE_NAMES.softNav;
const NO_LONG_TASK_WINDOW = exports.NO_LONG_TASK_WINDOW = 5000; // purpose is to wait 5 seconds wherein no long task is detected
const POPSTATE_MERGE_WINDOW = exports.POPSTATE_MERGE_WINDOW = 500; // "coalesce" (discard) a popstate that happen within this period following an INTERACTION_TRIGGER opening ixn, e.g. click->popstate

const INTERACTION_TYPE = exports.INTERACTION_TYPE = {
  INITIAL_PAGE_LOAD: '',
  ROUTE_CHANGE: 1,
  UNSPECIFIED: 2
};
const NODE_TYPE = exports.NODE_TYPE = {
  INTERACTION: 1,
  AJAX: 2,
  CUSTOM_END: 3,
  CUSTOM_TRACER: 4
};
const INTERACTION_STATUS = exports.INTERACTION_STATUS = {
  IP: 'in progress',
  PF: 'pending finish',
  // interaction meets the hard criteria but is awaiting flexible conditions to fully finish
  FIN: 'finished',
  CAN: 'cancelled'
};
