// @newrelic/browser-agent/dist/cjs/features/page_view_event/instrument/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PageViewEvent = exports.Instrument = void 0;
var _setPageViewName = require("../../../loaders/api/setPageViewName");
var _instrumentBase = require("../../utils/instrument-base");
var CONSTANTS = _interopRequireWildcard(require("../constants"));
var _constants2 = require("../../../common/session/constants");
var _globalEvent = require("../../../common/dispatch/global-event");
var _load = require("../../../common/window/load");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); } /**
 * Copyright 2020-2026 New Relic, Inc. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
class Instrument extends _instrumentBase.InstrumentBase {
  static featureName = CONSTANTS.FEATURE_NAME;
  constructor(agentRef) {
    super(agentRef, CONSTANTS.FEATURE_NAME);

    /** setup inspection events for window lifecycle */
    this.setupInspectionEvents();

    /** feature specific APIs */
    (0, _setPageViewName.setupSetPageViewNameAPI)(agentRef);
    this.importAggregator(agentRef, () => Promise.resolve().then(() => _interopRequireWildcard(require(/* webpackChunkName: "page_view_event-aggregate" */'../aggregate'))));
  }
  setupInspectionEvents() {
    const dispatch = (evt, name) => {
      if (!evt) return;
      (0, _globalEvent.dispatchGlobalEvent)({
        timeStamp: evt.timeStamp,
        loaded: evt.target.readyState === 'complete',
        type: 'window',
        name,
        data: evt.target.location + ''
      });
    };
    (0, _load.onDOMContentLoaded)(evt => {
      dispatch(evt, 'DOMContentLoaded');
    });
    (0, _load.onWindowLoad)(evt => {
      dispatch(evt, 'load');
    });
    (0, _load.onPopstateChange)(evt => {
      dispatch(evt, 'navigate');
    });
    this.ee.on(_constants2.SESSION_EVENTS.UPDATE, (_, data) => {
      (0, _globalEvent.dispatchGlobalEvent)({
        type: 'lifecycle',
        name: 'session',
        data
      });
    });
  }
}
exports.Instrument = Instrument;
const PageViewEvent = exports.PageViewEvent = Instrument;
