// @newrelic/browser-agent/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Agent", {
  enumerable: true,
  get: function () {
    return _agent.Agent;
  }
});
Object.defineProperty(exports, "Ajax", {
  enumerable: true,
  get: function () {
    return _ajax.Ajax;
  }
});
Object.defineProperty(exports, "BrowserAgent", {
  enumerable: true,
  get: function () {
    return _browserAgent.BrowserAgent;
  }
});
Object.defineProperty(exports, "GenericEvents", {
  enumerable: true,
  get: function () {
    return _generic_events.GenericEvents;
  }
});
Object.defineProperty(exports, "JSErrors", {
  enumerable: true,
  get: function () {
    return _jserrors.JSErrors;
  }
});
Object.defineProperty(exports, "Logging", {
  enumerable: true,
  get: function () {
    return _logging.Logging;
  }
});
Object.defineProperty(exports, "Metrics", {
  enumerable: true,
  get: function () {
    return _metrics.Metrics;
  }
});
Object.defineProperty(exports, "MicroAgent", {
  enumerable: true,
  get: function () {
    return _microAgent.MicroAgent;
  }
});
Object.defineProperty(exports, "PageAction", {
  enumerable: true,
  get: function () {
    return _page_action.PageAction;
  }
});
Object.defineProperty(exports, "PageViewEvent", {
  enumerable: true,
  get: function () {
    return _page_view_event.PageViewEvent;
  }
});
Object.defineProperty(exports, "PageViewTiming", {
  enumerable: true,
  get: function () {
    return _page_view_timing.PageViewTiming;
  }
});
Object.defineProperty(exports, "SessionReplay", {
  enumerable: true,
  get: function () {
    return _session_replay.SessionReplay;
  }
});
Object.defineProperty(exports, "SessionTrace", {
  enumerable: true,
  get: function () {
    return _session_trace.SessionTrace;
  }
});
Object.defineProperty(exports, "SoftNav", {
  enumerable: true,
  get: function () {
    return _soft_navigations.SoftNav;
  }
});
var _agent = require("./loaders/agent");
var _browserAgent = require("./loaders/browser-agent");
var _microAgent = require("./loaders/micro-agent");
var _ajax = require("./features/ajax");
var _jserrors = require("./features/jserrors");
var _generic_events = require("./features/generic_events");
var _logging = require("./features/logging");
var _metrics = require("./features/metrics");
var _page_action = require("./features/page_action");
var _page_view_event = require("./features/page_view_event");
var _page_view_timing = require("./features/page_view_timing");
var _session_trace = require("./features/session_trace");
var _session_replay = require("./features/session_replay");
var _soft_navigations = require("./features/soft_navigations");
