// @newrelic/browser-agent/dist/cjs/common/vitals/largest-contentful-paint.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.largestContentfulPaint = void 0;
var _attribution = require("web-vitals/attribution");
var _vitalMetric = require("./vital-metric");
var _constants = require("./constants");
var _runtime = require("../constants/runtime");
var _cleanUrl = require("../url/clean-url");
/**
 * Copyright 2020-2026 New Relic, Inc. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

const largestContentfulPaint = exports.largestContentfulPaint = new _vitalMetric.VitalMetric(_constants.VITAL_NAMES.LARGEST_CONTENTFUL_PAINT);
if (_runtime.isBrowserScope) {
  (0, _attribution.onLCP)(({
    value,
    attribution
  }) => {
    /* Largest Contentful Paint - As of WV v3, it still imperfectly tries to detect document vis state asap and isn't supposed to report if page starts hidden. */
    if (_runtime.initiallyHidden || largestContentfulPaint.isValid) return;
    let attrs = {
      timeToFirstByte: attribution.timeToFirstByte,
      resourceLoadDelay: attribution.resourceLoadDelay,
      resourceLoadDuration: attribution.resourceLoadDuration,
      resourceLoadTime: attribution.resourceLoadDuration,
      // kept for NR backwards compatibility, deprecated in v3->v4
      elementRenderDelay: attribution.elementRenderDelay
    };
    const lcpEntry = attribution.lcpEntry;
    if (lcpEntry) {
      attrs.size = lcpEntry.size;
      attrs.eid = lcpEntry.id;
      if (lcpEntry.element?.tagName) attrs.elTag = lcpEntry.element.tagName;
    }
    if (attribution.element) attrs.element = attribution.element;
    if (attribution.url) attrs.elUrl = (0, _cleanUrl.cleanURL)(attribution.url);
    largestContentfulPaint.update({
      value,
      attrs
    });
  });
}
